samcli/commands/remote_invoke/remote_invoke_context.py

```diff
@@ -0,0 +1,150 @@
+from samcli.lib.utils.boto_utils import BotoProviderType
+from samcli.lib.utils.cloudformation import (
+    CloudFormationResourceSummary,
+    get_resource_summaries,
+    get_resource_summary_from_physical_id,
+)
+from samcli.lib.utils.resources import AWS_LAMBDA_FUNCTION
+
+LOG = logging.getLogger(__name__)
+
+
+SUPPORTED_SERVICES = {"lambda": AWS_LAMBDA_FUNCTION}
+
+
+class RemoteInvokeContext:
+
+    _boto_client_provider: BotoProviderType
+    _boto_resource_provider: BotoProviderType
+    _stack_name: Optional[str]
+    _resource_id: Optional[str]
+    _resource_summary: Optional[CloudFormationResourceSummary]
+
+    def __init__(
+        self,
+        boto_client_provider: BotoProviderType,
+        boto_resource_provider: BotoProviderType,
+        stack_name: Optional[str],
+        resource_id: Optional[str],
+    ):
+        self._boto_resource_provider = boto_resource_provider
+        self._boto_client_provider = boto_client_provider
+        self._stack_name = stack_name
+        self._resource_id = resource_id
+        self._resource_summary = None
+
+    def __enter__(self) -> "RemoteInvokeContext":
+        self._populate_resource_summary()
+        return self
+
+    def __exit__(self, *args) -> None:
+        pass
+
+    def run(self):
+        pass
+
+    def _populate_resource_summary(self) -> None:
+        """
+        Populates self._resource_summary field from self._stack_name and/or self._resource_id
+
+        Either self._stack_name or self._resource_id should be defined, it fails otherwise.
+
+        If only self._stack_name is defined, it tries to find single resource in that stack,
+        see _get_single_resource_from_stack for details.
+
+        If only self._resource_id is defined, it tries to parse its ARN or validate it as physical id,
+        see _get_from_physical_resource_id for details.
+        """
+        if not self._stack_name and not self._resource_id:
+            raise InvalidRemoteInvokeParameters("Either --stack-name or --resource-id parameter should be provided")
+
+        if not self._resource_id:
+            # no resource id provided, list all resources from stack and try to find one
+            self._resource_summary = self._get_single_resource_from_stack()
+            self._resource_id = self._resource_summary.logical_resource_id
+
+        if not self._stack_name:
+            # no stack name provided, resource id should be physical id so that we can use it
+            self._resource_summary = self._get_from_physical_resource_id()
+
+    def _get_single_resource_from_stack(self) -> CloudFormationResourceSummary:
+        """
+        Queries all resources from stack with its type,
+        and returns its information if stack has only one resource from that type (including nested stacks)
+        """
+        LOG.debug(
+            "Trying to get single resource with %s type in % stack since no resource id is provided",
+            AWS_LAMBDA_FUNCTION,
```

Comment:
Should we change this to `SUPPORTED_SERVICES.values()` as well? When we add support for more services, we can have a priority of which resource gets invoked when there are single resources of different services.